sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -80,462 +65,182 @@ protected sealed override void Initialize(SonarAnalysisContext context)
+
+                            foreach (var symbol in allNamedTypes)
+                            {
+                                VisitDeclaringReferences(symbol, usageCollector);
+                            }
+
+                            var diagnostics = GetDiagnostics(usageCollector, removableInternalMembers.ToHashSet(), "internal",
+                                new BidirectionalDictionary<ISymbol, SyntaxNode>());
+                            foreach (var diagnostic in diagnostics)
+                            {
+                                cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
+                            }
+                        });
                 });
         }
 
-        private static bool IsInternalVisibleToAttribute(AttributeSyntax attribute, SemanticModel semanticModel)
+        private static IEnumerable<Diagnostic> GetDiagnostics(SymbolUsageCollector usageCollector, ISet<ISymbol> removableSymbols,
+            string accessibility, BidirectionalDictionary<ISymbol, SyntaxNode> fieldLikeSymbols)
         {
+            var unusedSymbols = removableSymbols
+                .Except(usageCollector.UsedSymbols)
+                .Where(symbol => !MentionedInDebuggerDisplay(symbol))
+                .ToHashSet();
+
+            var propertiesWithUnusedAccessor = removableSymbols
+                .Intersect(usageCollector.UsedSymbols)
+                .OfType<IPropertySymbol>()
+                .Where(usageCollector.PropertyAccess.ContainsKey);
 
-            return semanticModel.GetSymbolInfo(attribute).Symbol is IMethodSymbol attributeConstructor &&
-                attributeConstructor.ContainingType.Is(KnownType.System_Runtime_CompilerServices_InternalsVisibleToAttribute);
+            // Report private unused symbols
+            return GetDiagnosticsForMembers(unusedSymbols, accessibility, fieldLikeSymbols)
+                .Concat(GetDiagnosticsForPropertyAccessors(propertiesWithUnusedAccessor, usageCollector.PropertyAccess));
+
+            bool MentionedInDebuggerDisplay(ISymbol symbol) =>
+                usageCollector.DebuggerDisplayValues.Any(value => value.Contains(symbol.Name));
         }
 
-        private static void ReportUnusedPropertyAccessors(SymbolAnalysisContext context, HashSet<ISymbol> usedSymbols,
-            HashSet<ISymbol> declaredPrivateSymbols,
+        private static IEnumerable<Diagnostic> GetDiagnosticsForPropertyAccessors(IEnumerable<IPropertySymbol> onlyOneAccessorAccessed,
             Dictionary<IPropertySymbol, AccessorAccess> propertyAccessorAccess)
         {
-            var usedPrivatePropertis = declaredPrivateSymbols.Intersect(usedSymbols).OfType<IPropertySymbol>();
-            var onlyOneAccessorAccessed = usedPrivatePropertis.Where(p => propertyAccessorAccess.ContainsKey(p));
+            return onlyOneAccessorAccessed.SelectMany(GetDiagnosticsForProperty).ToList();

```

Comment:
I find this style of using inner functions much harder to read and understand (particularly here, where the method is one line, but contains a twenty line inner function and another inner function after that. I have a similar problem with the  _GetDiagnosticsForMembers_ method; there's a _return_ statement half-way down the method, and everything after that is inner functions.

I'm not sure what a good style guideline would be. Perhaps limiting local functions to things that can be expressed in one or two lines using expression syntax?

Reply:
+1, it looks better with less nested methods.